### ISSUE TITLE
Improve typing

### ISSUE THREAD

**Issue by mrossinek** [qiskit-community/qiskit-algorithms#74]
### What should we add?

While working on #73, @ElePT and I have identified the following inconsistencies and/or problems around the type hints which should be investigated further. Rather than holding up that PR unnecessarily, we decided to gather these here.
That way, they can be investigated one-by-one.

Note: some of these may very likely require code changes to resolve inconsistencies in the interfaces.

- [ ] Typing around `np.ndarray` vs `Sequence[float]`. It is a known issue that to `mypy` the numpy arrays do not satisfy the `Sequence` type. This leads to many type ignore statements right now which can hopefully be improved upon.
  - [ ] For example, typing around parameter values being passed on to the primitives: sometimes its `Sequence[float]`, sometimes its `Sequence[Sequence[float]]`, and oftentimes internally it ends up as `np.ndarray` causing many type ignore comments
- [ ] The typing of the `initial_point` attributes is very inconsistent across its various occurrences.
- [ ] The `Minimizer` interface does not appear to be supported properly by `VQD` and `PVQD` (for example) since they do not provide the required `jac` argument
- [ ] The `ListOrDict` type is causing quite some oddities. This has multiple cases:
    - [ ] for the `metadata` (e.g. `variance`) types and **implementations** are inconsistent with the variance sometimes wrapped in dictionaries and sometimes provided as a single number
    - [ ] for `aux_operators` and/or `observables` we had to add many type ignores which would need to be re-evaluated (possibly linked to the previous point)
- [ ] All classes that use the `EsimationProblem` have very poor typing around the following points:
    - [ ] the `post_processing` function
    - [ ] the confidence intervals
    - [ ] the `circuit_results`

**Comment by woodsp-ibm** [comment]
The Estimation logic - and possible elsewhere - has code that is no longer used. While deprecated code for QI/opflow was removed there are some util like functions  taking types that would have been an outcome - this code is now untested.

Also Grover has some issue around using the circuit passed a as grover operator - it allows QuatumCircuit but then treats it as GroverOperator where it unconditionally accesses an attribute that has which is not present on a QuantumCircuit.

This was just from starting to look at mypy in AA and AE folders.

**Comment by mrossinek** [comment]
I suppose a good strategy at finding which of the code is unused would be to look at the coverage reports. Obviously not all uncovered lines can be removed but it might be a good starting point :+1:

**Comment by woodsp-ibm** [comment]
Yes, particularly around the algos AE, Grover and Phase Estimation that were modified to add Sampler logic paths to the existing code and whether more could/should have been removed when QI paths were removed that was part of common utilities.
